docs(rfc): violation liveness and resolution tracking (#678)
rancher-sandbox/runtime-enforcer: docs/rfc/0008-violation-liveness.md

**Executable added to the allowlist.** A violation for `/usr/local/bin/debug-tool`
sits in status with `active: true`. The operator edits the policy and adds the
path under `spec.rulesByContainer["app"].executables.allowed`. On the next
reconcile the controller sees the match and flips the record to
`active: false`. Removing the entry from the allowlist reactivates it.

**Workload redeployed.** A violation references pod `app-7d8f9c-x2k4p` which
ran a binary dropped at runtime into `/tmp`. The operator does
`kubectl rollout restart deployment/app`. The old pod is deleted, a new one
comes up with a different name. The controller's pod informer no longer finds
`app-7d8f9c-x2k4p` in the cluster, so the violation becomes `active: false`.
The new pod starts with no violations of its own.

**Mixed state.** After a few days a policy might carry:

```
/usr/local/bin/debug-tool   active: false   (allowlisted)
/tmp/malware.sh             active: false   (pod gone)
/opt/scripts/cleanup.sh     active: true    (still running, not allowed)
/usr/bin/nc                 active: true    (still running, not allowed)
```

Filtering on `active == true` gives you the two items that actually need

> attention.

Reviewer: I don't think that marking a resolved the issues that came from a pod that is no longer running is a good idea.

Take the example you made, the `malware.sh` that was blocked now is no longer standing out. IMHO it should be visible for as long as a user decides to silence it.

Reviewer 2: the pod UID should remain stable unless there is a rollout so we won't silence a warning unless the parent workload (deployment, daemonset) is deleted or there is a rollout. If after malware.sh the container restarts the pod UID doesn't change so we shouldn't resolve the violation.
Unfortunately Job/Cronjobs are kind of a corner cases because if there is a violation inside them the and they complete their task, the violation will be automatically solved

# Detailed design

[design]: #detailed-design

The `ViolationRecord` struct gains one field:

```go
// active indicates whether the violation is currently live.
// +optional

> Active bool `json:"active,omitempty"`

Reviewer 2: instead of `Active` why don't we change the name of the violation array to `ActiveViolations` and when a violation is no more active we drop it or we move it to another array `ResolvedViolation`? What i don't love is the fact that the violation could change state multiple times while once resolved the violation should stay resolved

Reviewer 2: * let's say we have a violation on  `/usr/bin/curl` 
* the user will add curl because he forgot about it. 

IMO the violation should be resolved and it shouldn't have the possibility to become active again by design. Keeping an active field gives the wrong idea that the violation could become active again, but i don't think this should be the case

Author: What I thought initially was to have a list of `violationExceptions` or something like that, but the concept is very much the same. I agree that an expired violation should stay expired, I'm just trying to wrap my head around this without second guessing myself

Reviewer: I like this idea, let's keep that in mind

conditions holds:

1. **The binary is now allowlisted.** For the violation's `containerName`, the
   controller scans `spec.rulesByContainer[containerName].executables.allowed`
   and looks for a match on `executablePath`. This covers the case where the
   operator has explicitly accepted the binary as part of the policy.

2. **The pod is gone.** The controller runs a pod informer and looks up the
   violation's `podName` in the cache. If the pod no longer exists in the
   cluster, the violation is expired. This covers the case where the workload
   has been redeployed (new image, `rollout restart`, node drain, etc.) and
   the offending pod instance, along with anything dropped into its writable
   layer, no longer exists.

If neither condition holds, the violation stays `active: true`. The controller
recomputes the field on every reconcile for every record in the list.

The proto, the agent's `ViolationBuffer`, and the OTEL event pipeline are
untouched. RBAC is already sufficient: the controller has `get/list/watch`
on pods. On upgrade, existing records without the field default to active
and get re-evaluated on the first reconcile.

# Drawbacks

[drawbacks]: #drawbacks

> - Pod name collisions on StatefulSets can briefly mis-report liveness: a new

Reviewer 2: uhm this is definitely not ideal, maybe we should add the podUID in the violation record, so that we are sure when a pod is restarted

Author: I agree :+1:

  pod with the same name makes an old record look active again. The
  100-record trim and the fact that a genuinely recurring violation will
  produce a fresh record with a new timestamp make this tolerable for now.

> - Runtime-dropped binaries are covered eventually (pod is gone, violation expires).

Reviewer 2: to be honest this is not super clear to me. If the image is changed the violations should expire since the pod will be redeployed, am i missing something?

Author: Yes but for example if you change the image and you keep having an entrypoint that does `wget` of something and that binary is nasty you can incur in new violations. I can strip this off, I just included it for the sake of completeness :smile:

  Image-resident binaries are only covered via the allowlist path. If a bad
  binary ships in the image and nobody updates the allowlist, the violation
  stays active across pod restarts of the same image. That is the correct
  behavior, but it does mean the image rebuild path to resolution is not
  automated; see the rejected alternative below for why.

# Alternatives considered

[alternatives]: #alternatives

## Rejected: violation liveness via image digest

We also considered a design where each `ViolationRecord` would carry the
container's resolved image **digest** (from

> `.status.containerStatuses[].imageID`, never the mutable tag), captured at

Reviewer: While having the image digest is better, this makes things more complicated (as pointed out next).

What if we stick to the image tag instead? We can enforce best practices (using shasums) with documentation and with a mutating admission controller (KW adm controller has already a policy for that).

context about which workload produced each violation. We enrich violation
records with workload metadata, add explicit violation acknowledgements to the
policy spec, and delete expired entries when building the violations list.
This gives operators meaningful context and a clean way to acknowledge known
issues while automatically removing resolved ones.

# Motivation

[motivation]: #motivation

Without liveness, every violation looks equally urgent. A security engineer
triaging a policy with 40 violations has no way to know that 35 of them came
from pods that were redeployed last week, or that 3 more refer to binaries
that have since been allowlisted. Dashboards, alerts, and compliance reports
all end up noisier than they need to be, and there is no clean signal for
"is this workload currently clean?"

## Examples

[examples]: #examples

**Executable added to the allowlist.** A violation for `/usr/local/bin/debug-tool`
sits in the violations list. The operator edits the policy and adds the
path under `spec.rulesByContainer["app"].executables.allowed`. On the next
reconcile the controller sees the match and deletes the record from the list.

> **Workload redeployed.** A violation references pod `app-7d8f9c-x2k4p` which

Reviewer 2: if i recall correctly we don't need anymore this case, there should be only 2 cases:
* the user forgot a binary in the policy -> the user adds it manually and the controller will clean the violations
* there is a real violation -> the user needs to `acknowledged` it and the controller will clean the acknowledged violation.

Reviewer: yes, I agree with @Andreagit97

Author: Sorry, leftover lol

ran a binary dropped at runtime into `/tmp`. The operator does
`kubectl rollout restart deployment/app`. The old pod is deleted, a new one
comes up with a different name. The controller's pod informer no longer finds
`app-7d8f9c-x2k4p` in the cluster, so the violation is deleted from the list.
The new pod starts with no violations of its own.

**Violation acknowledged.** An operator sees a violation for `/opt/scripts/cleanup.sh`
and knows it's part of an ongoing incident. They add it to

> `spec.acknowledgedViolations` with a reason and match criteria. On the next

Reviewer: I'm not convinced we should add this information to the `spec` field, since it not about declaring a desired state.

Author: Sounds reasonable. Maybe in the `status`?

Reviewer 2: uhm the status doesn't seem the correct place, since it is something controller-owned and shouldn't be touched by the user. The user here is asking the controller to acknowledge the violations so it doesn't seem wrong to put it into the spec, we can see it as a desired state "we want these violations to be acknowledged", and the controller will eventually acknowledge them

Author: My two cents: status is out, spec works, the correct answer to this would be a separate resource but I also want to prevent premature CRD proliferation and keep a KISS approach. Folding it into the policy spec means one RBAC grant covers both triage and enforcement, which we'll regret in a security product. I'm down to this but let's keep it self-contained and document the RBAC coupling as known debt.

Reviewer: Sorry, I really don't like the usage of `status.something` for that.

What about having an annotation per acknowledged incident?

```yaml
apiVersion: runtime-enforcer.io/v1alpha1
kind: Policy
metadata:
  name: restrict-app-exec
  annotations:
    # Pattern: [prefix]/[violation-id] : [reason]
    ack.runtime-enforcer.io/b2e45a90: "Dev team investigating memory leak, allowed temporary exec"
    ack.runtime-enforcer.io/c8f19b22: "Ops ticket INC-9942 - manual DB migration script"
spec:
```

reconcile, the controller deletes it from `status.violations`, leaving only
unacknowledged violations visible.

**Mixed state.** After a few days, only unresolved violations remain in the list:

```
/opt/scripts/cleanup.sh     (acknowledged, deleted from list)
/usr/bin/nc                 (still running, not allowed, still in list)
```

The violations list now contains only items that actually need attention.

# Detailed design

[design]: #detailed-design

Each `ViolationRecord` gains workload fields (`workloadName`, `workloadKind`,

> `workloadNamespace`) alongside the existing `podName`, giving operators

Reviewer: I think we can avoid the workloadNamespace, since the policy sits exactly in the same namespace as the workload.

immediate context about which workload produced the violation. This is
populated from the pod's owner reference when the violation is recorded.

> The policy spec gains an `acknowledgedViolations` list where operators can

Reviewer: Maybe make that a top level key of the object? I would do a quick search to see if there are some best practices for that.

Author: Like the id field but generated? I like it

explicitly mark specific violations as known and accepted. Each entry includes
an identifier, a reason string, and match criteria (pod UID, container name,
executable path). For example:

```yaml
spec:
  acknowledgedViolations:
  - id: incident-1--2026-06-12

>     reason: "incident in triage"

Reviewer: The user has to come up with `id` and `reason`. I'm fine with reason, I wonder about id, we might want to help the user come up with an automatic value for that field

    match:
      podUID: "8f3c0e3d-..."
      containerName: "app"
      executablePath: "/tmp/malware.sh"
```

When the controller reconciles and builds the violations list, it deletes

> entries that are expired (pod no longer exists, or executable is now

Reviewer 2: ```suggestion
entries that are expired (executable is now
```

allowlisted) or acknowledged (matches an entry in `spec.acknowledgedViolations`).
This keeps `status.violations` clean and actionable without requiring operators
to filter on status fields.

The agent and gRPC proto gain workload fields to pass through to the

> controller. RBAC is already sufficient: the controller has `get/list/watch`

Reviewer 2: the agent should already provide the information on the workload through gRPC so the controller shouldn't do anything.

on pods and can read owner references to populate workload metadata. On
upgrade, existing records without workload fields are left as-is and still
subject to expiration checks.

> This design prioritizes clarity over history preservation: resolved violations

Reviewer: please extend the RFC by pointing out that for history purpose the user must configure an endpoint to receive all the otel logs we generate

disappear rather than accumulating with status markers. The allowlist and
acknowledgements list in `spec` provide the audit trail for why violations
were cleared.

# Drawbacks

[drawbacks]: #drawbacks

- Deleting expired violations means we lose historical data about past
  incidents. Operators who want to audit what happened last month will need
  to rely on external logging or the OTEL event pipeline rather than the

>   policy's status field.

Reviewer: I would move that above

- Pod name collisions on StatefulSets can cause premature deletion: if a new
  pod with the same name comes up before the violation is reconciled, the
  old violation might persist longer than intended. Using pod UIDs in the

>   match criteria mitigates this for acknowledged violations.

Reviewer: good idea, can we make pod UID optional? we could have our kubectl plugin deal with that automatically

Author: Absolutely!

allowlisted. While this preserved history and avoided mutation of the
violations list, it made every violation look equally important in dashboards
and required operators to filter on `active == true` to find actionable items.
The approach also had no way for operators to explicitly acknowledge a
violation as accepted during triage, leaving no audit trail for "I know about
this and it's okay." Deletion of resolved violations, combined with explicit
acknowledgements in the spec, provides a cleaner signal and better matches how
security teams actually work.

## Rejected: violation liveness via image digest

We also considered a design where each `ViolationRecord` would carry the
container's resolved image **digest** (from
`.status.containerStatuses[].imageID`, never the mutable tag), captured at
the time the violation was emitted by the agent. A record would be active
if and only if both its digest matched what the container is currently
running and its `executablePath` was not in the container's allowlist. The
controller would expose a derived `activeViolationCount` and a `Compliant`
condition (`True` when the count is zero), giving monitors a single field
to read.

> Clearance would be tied to the actualremediation action: rebuilding the image

Reviewer 2: ```suggestion
Clearance would be tied to the actual remediation action: rebuilding the image
```

  convention agreed between agent and controller.
- Most damaging: it only covers image-resident executables. Binaries
  dropped at runtime into a writable layer or tmpfs have the lifetime of
  the pod instance, not the image. Closing that gap would require
  executable-origin detection at the agent, which does not exist today, so
  the added machinery would not on its own fully solve clearance.

We should revisit this if the simpler approach starts flapping or producing
false-healthy signals in practice, or once the agent can tell image-resident
binaries from runtime-dropped ones. At that point the two mechanisms could
be combined: digest liveness for image-resident violations, pod-instance
liveness for runtime-dropped ones, which would give correct derived
recovery across both classes and justify the complexity.

# Unresolved questions

[unresolved]: #unresolved-questions

- Should we add deduplication of violation records by (workload, container,
  executable) with a count field? Repeated attacks can produce many similar
  records, potentially causing unwanted data loss when the 100-record cap is
  hit. This is orthogonal to liveness and can be added separately.

> - Should the controller expose a derived `activeViolationCount` in status,

Reviewer 2: yep this seems a good thing to have

referencing it:

```yaml
metadata:
  annotations:
    runtime-enforcer.security.rancher.io/acknowledge/9f1c2b3e-...: "ongoing incident, see JIRA-1234"
```

On the next reconcile the controller:

1. Finds the annotation, extracts the violation `id` and `reason`.
2. Looks up the matching record in `status.violations` and removes it.
3. Appends it to `status.acknowledgedViolations` (carrying the original
   record, the reason, and the acknowledgement timestamp).
4. Emits an OTEL log with the violation data and the reason.
5. Removes the annotation from the policy.

After this, `status.violations` only contains items that actually need
attention, and `status.acknowledgedViolations` is the audit trail of
explicitly triaged ones.

**Mixed state.** After a few days, only unresolved violations remain in the
list:

```
status.violations:
  /usr/bin/nc                 (still running, not allowed, still in list)

status.acknowledgedViolations:
  /opt/scripts/cleanup.sh     (acknowledged via annotation, moved here)
```

The violations list now contains only items that actually need attention.

# Detailed design

[design]: #detailed-design

## Violation record enrichment

> Each `ViolationRecord` gains a stable `id` (a UUID generated by the controller

Reviewer 2: do you see any advantage in using a specific generated UUID over the incremental counter that we already have in the policy?

Author: Just going with a uint64 is ok for now, thanks

## Acknowledgement via annotation

Operators acknowledge a specific violation by setting an annotation on the
`WorkloadPolicy` with the form:

```
<annotation-prefix>/<violation-id>: "<reason>"
```

The default annotation prefix is
`runtime-enforcer.security.rancher.io/acknowledge`. The key's suffix is
the `id` of the violation being acknowledged, taken from
`status.violations[].id`. The value is a free-form human-readable reason
recorded into `status.acknowledgedViolations`.

Example:

```yaml
metadata:
  annotations:
    runtime-enforcer.security.rancher.io/acknowledge/9f1c2b3e-4d5a-6b7c-8d9e-0f1a2b3c4d5e: "ongoing incident, see JIRA-1234"
```

On every reconcile the controller:

1. Collects all annotations on the policy whose key has the acknowledgement
   prefix.
2. For each such annotation, parses the violation `id` from the key and the
   `reason` from the value.
3. Looks up the record in `status.violations` by `id`. If found:
   - Removes it from `status.violations`.
   - Appends a new entry to `status.acknowledgedViolations` carrying the
     full original `ViolationRecord`, the `reason`, and the
     acknowledgement timestamp.

>    - Emits an OTEL log record with `event_name=policy_violation_acknowledged`

Reviewer 2: not sure if we want the otel log on the acknowledgement in the first iteration

     containing the violation data and the reason (see OTEL logging below).
   - Records the annotation key in a "to-be-removed" set.
4. If the `id` does not match any current violation, the annotation is

>    treated as a no-op (still removed), and a warning OTEL log is emitted

Reviewer 2: maybe we can start with a warning in the controller

   so the operator can tell their annotation was stale.
5. After processing all acknowledgement annotations, the controller removes
   them from the policy object via a patch. Acknowledgements are
   one-shot: once consumed, the annotation is gone and the acknowledgement
   lives on in `status.acknowledgedViolations`.

The acknowledgement is "punctual" in the sense that it targets one specific
violation by `id` and disappears after being processed. It does not affect
future violations of the same executable, so a reappearing violation will
have to be acknowledged again.

> 

Reviewer: We also have to mention the `status.acknowledgedViolations` is capped to 100 items

  fast with a clear error message (so the operator gets immediate
  feedback rather than a silent no-op reconcile).
- Patches the policy metadata to add the acknowledgement annotation
  with the resolved `<reason>` (prompted interactively if not passed
  via `--reason`).
- Supports `--dry-run` to print the patched object without sending it
  to the API server, matching the existing `allow`/`deny` subcommands.
- Supports shell completion for `<violation-id>` from
  `status.violations[].id` of the target policy, similar to how the
  `allow` subcommand completes from the violations list.

The subcommand is a thin wrapper: it does not know or care how the
controller reconciles the annotation, it only writes the annotation. The
controller is still the single source of truth for the
`acknowledgedViolations` list and the OTEL log. This keeps the
annotation-based design layered cleanly: the controller speaks
annotations, the plugin speaks operator UX, the two are kept in sync by
the annotation contract.

## RBAC and upgrade

The agent and gRPC proto will pass the workload provenance data to the
controller. RBAC is already sufficient: the controller has `get/list/watch`

> on pods and can read owner references to populate workload metadata. The

Reviewer 2: This is not true
```suggestion
controller. The
```

controller needs `update`/`patch` on `workloadpolicies` to remove the
acknowledgement annotations, which it already has (it writes status today).

> On upgrade, existing records without `id` and workload fields are

Reviewer 2: i don't think this section is relevant since we are not yet running this software in production

agent on a later tick is the same one it already knows about, so the `id`
(and workload fields) are preserved across scrapes.

## Violation `id` as an evergrowing integer

> The `id` is a per-policy, monotonically increasing `int64` assigned by the

Reviewer: we should make that a uint64 to have even more room for growth
